Non-mandatory fuel

v0d1ch opened the pull request:
* Deprecates the use of internal commits and the need for marking fuel
* Adds some documentation about external commits
* Updates documentation such that marking utxo as fuel is not needed anymore
* Makes end-to-end tests and benchmarks use the external commit
  - Keeps one test for the internal commit to verify it still works
* Uses largest UTxO owned by `--cardano-signing-key` if no marked fuel can be found.

---


* [x] CHANGELOG updated or not needed
* [x] Documentation updated or not needed
* [x] Haddocks updated or not needed
* [x] No new TODOs introduced or explained herafter

v0d1ch committed:
Removals and renames

- Remove NoFuelUTxOFound
- rename NotEnoughFuel to NotEnoughFunds
- rename findFuelUTxO to findLargestUTxO and make sure we provide
  appropriate error message.

v0d1ch committed:
Update changelog

v0d1ch committed:
Depreciate fuel and add a section about external commits in quickstart

v0d1ch committed:
Add hydra client function to draft external commit tx

We needed to move functions to avoid cyclic dependency in order
to define submitTx using RunningNode

ch1bo committed:
Replace draft tx submission in external commit e2e test

Re-using the 'externalCommit' function here hints at the fact that now
every e2e test with a commit will be doing just the same what was tested here.

ch1bo committed:
Use externalCommit in an e2e tests which commits something

This also adds the signTx function to conveniently sign already prepared
transactions.

v0d1ch committed:
Wip

ch1bo committed:
Generate benchmark data sets with external keys

We need to fund not the internal, but another key if we want to commit
from external now.

ch1bo committed:
Only fund the external signing keys in the funding transaction

This is enough as the benchmark does fund the internal signing keys of
each hydra-node at the beginning of the scenario.

ch1bo committed:
Add TODOs for inverting how we treat fuel

v0d1ch committed:
Convert one e2e test to commit utxo using internal wallet

v0d1ch committed:
Try to find the fueled utxo but fallback to using the largest one

v0d1ch committed:
Add unit test for findFuelOrLargestUTxO

v0d1ch committed:
Rebase master

v0d1ch committed:
Use Req library

ffakenz committed:
Move signTx to hydra-cardano-api

ffakenz commented on hydra-cluster/src/CardanoNode.hs:
```diff
@@ -188,7 +186,6 @@ withCardanoNodeOnKnownNetwork tracer workDir knownNetwork action = do
   networkId <- readNetworkId
   copyKnownNetworkFiles
   withCardanoNode tracer networkId workDir args $ \node -> do
-    waitForFullySynchronized tracer node
```
why this got removed?

ch1bo replied:
It just got moved, see `withCardanoNodeOnKnownNetwork` call sites. Reason was to avoid cyclic module dependency.

ffakenz commented on hydra-node/test/Hydra/Chain/Direct/WalletSpec.hs:
```diff
@@ -103,6 +105,9 @@ spec = parallel $ do
         assertQueryPoint QueryTip
 
+    prop "prefers fueled utxo to pay the fees" prop_picksCorrectUTxOToPayTheFees
+    prop "prefers largest utxo if no fuel available" prop_picksLargestUTxOToPayTheFees
```
Nice job with this properties! 👏

ffakenz commented on hydra-node/src/Hydra/Chain/Direct/Wallet.hs:
```diff
@@ -365,16 +361,29 @@ coverFee_ pparams systemStart epochInfo lookupUTxO walletUTxO partialTx@Babbage.
+      case UTxO.pairs $ UTxO.maxLovelaceUTxO apiUtxo of
         [] -> Nothing
-        as -> Just (List.last as)
+        as ->
```
```suggestion
        highestValueUtxo ->
```

ffakenz commented on hydra-node/src/Hydra/Chain/Direct/Wallet.hs:
```diff
@@ -365,16 +361,29 @@ coverFee_ pparams systemStart epochInfo lookupUTxO walletUTxO partialTx@Babbage.
-      sortingCriteria (_, Babbage.BabbageTxOut _ v _ _) = fst' (gettriples' v)
-      sortedByValue = sortOn sortingCriteria utxosWithDatum
-   in case sortedByValue of
+findFuelOrLargestUTxO :: Map TxIn TxOut -> Maybe (TxIn, TxOut)
```
Do you think we should add a note for when deprecating the Fuel?

ffakenz committed:
Fix docs links to external-commits to be relative

Also:
- Improve changelog.
- Enhance fuel warnings in docs.

ch1bo commented on docs/docs/getting-started/quickstart.md:
```diff
@@ -203,8 +203,18 @@ We provide existing files in [hydra-cluster/config](https://github.com/input-out
+
+Hydra node allows the users to commit utxo using their own wallet. This removes the need to mark some specific utxo as _fuel_ to drive L1 transactions. Now it is the users obligation to provide enough funds for a key specified with the `--cardano-signing-key` hydra-node flag since this key is used to pay for all L1 transactions.
+
+The node provides the http endpoint at the `/commit` path of the internal api server which can accept multiple user utxos (belonging to public key or script address) to commit to a Head after the `Initialization` phase. Hydra-node sends back a draft transaction which is expected to be submitted to a network by the user. Please take a look at the [api documentation](https://hydra.family/head-proto ...
```
The information here is correct, but don't feel afraid in changing/incorporating this more with the `Fuel` section below!

For example, the explanation for fuel per se is still good to have, only difference is that now it does not _need_ to be marked if we use external commits.

Also, there is a an `:::info About commits` banner below which feels odd to have if we have this section.

Should: update the `Fuel` section, and maybe move this after it, only adding additional info

ch1bo replied:
Done in https://github.com/input-output-hk/hydra/commit/241389ae368f19421d2b7ca29bae938720b4a1a6

ch1bo commented on docs/docs/tutorial/using_hydra/using-hydra-part-2.md:
```diff
@@ -69,6 +69,10 @@ cardano-cli query utxo --address $(cat ./test-head/Alice/AliceCardano.addr) --te
 
+:::warning Fuel is deprecated and will be removed in future Hydra versions.
+Please take a look at [external-commits](/head-protocol/docs/getting-started/quickstart#external-commits)
+:::
```
In a perfect world we would have updated the tutorial already with external commits. But it may be outdated for other reasons now.

Let's keep this part in mind when we write the follow-up of "actually removing fuel"

ch1bo commented on hydra-cardano-api/src/Cardano/Api/UTxO.hs:
```diff
@@ -80,6 +80,18 @@ render (k, TxOut _ (txOutValueToValue -> v) _ _) =
+    [] -> error "Empty UTxO has no maximum value."
+    as -> fromPairs [List.head $ List.sortOn sortingCriteria as]
+ where
+  sortingCriteria (_, TxOut _ v _ _) = txOutValueToLovelace v
```
Must: not write partial functions, especially not in library code

This is an oddly specific function and is likely only used in our wallet code. Let's keep it there and write it in a non-partial way?

How to improve the function directly: The return type should be a single entry or none, so I'd suggest `Maybe (TxIn, TxOut)` to make it total.

ch1bo commented on hydra-cardano-api/src/Cardano/Api/UTxO.hs:
```diff
@@ -80,6 +80,18 @@ render (k, TxOut _ (txOutValueToValue -> v) _ _) =
+-- NOTE: This function is partial.
+maxLovelaceUTxO :: UTxO -> UTxO
+maxLovelaceUTxO utxo =
+  case Map.toList $ toMap utxo of
```
Should: use `UTxO.pairs`

ch1bo commented on hydra-cardano-api/src/Cardano/Api/UTxO.hs:
```diff
@@ -80,6 +80,18 @@ render (k, TxOut _ (txOutValueToValue -> v) _ _) =
+maxLovelaceUTxO utxo =
+  case Map.toList $ toMap utxo of
+    [] -> error "Empty UTxO has no maximum value."
+    as -> fromPairs [List.head $ List.sortOn sortingCriteria as]
```
Could: use `sortOn (txOutValueToLovelace . snd)`

ch1bo commented on hydra-cardano-api/src/Hydra/Cardano/Api.hs:
```diff
@@ -706,3 +706,15 @@ pattern ScriptWitness scriptWitnessInCtx scriptWitness <-
+signTx signingKey (Tx body wits) =
+  makeSignedTransaction (witness : wits) body
+ where
+  witness = makeShelleyKeyWitness body (WitnessPaymentKey signingKey)
```
Should: follow the organization guidelines in the `hydra-cardano-api/README.md` and move this to `Hydra.Cardano.Api.Tx` module, with a re-export in this module here.

ch1bo replied:
https://github.com/input-output-hk/hydra/commit/b62f839a1e70998e04f059eba6050e5a0005131a

ch1bo commented on hydra-cluster/src/Hydra/Cluster/Scenarios.hs:
```diff
@@ -143,49 +139,32 @@ singlePartyHeadFullLifeCycle tracer workDir node@RunningNode{networkId} hydraScr
     traceWith tracer RemainingFunds{actor = actorName actor, fuelUTxO, otherUTxO}
 
-singlePartyCommitsFromExternal ::
+-- Make sure the _old_ way of committing (using Fuel) still works.
```
Should: prefer haddock comments on function signatures

```suggestion
-- | Ensures the _old_ way of committing (using Fuel) still works.
```

ch1bo commented on hydra-cluster/test/Test/EndToEndSpec.hs:
```diff
@@ -132,11 +133,11 @@ spec = around showLogsOnFailure $ do
               >>= timedTx tmpDir tracer node
-      it "commits from external wallet" $ \tracer -> do
-        withClusterTempDir "single-commits-from-external" $ \tmpDir -> do
+      it "commits using fuel" $ \tracer -> do
```
:+1:

ch1bo commented on hydra-cluster/src/Hydra/Generator.hs:
```diff
@@ -86,39 +94,35 @@ genDatasetConstantUTxO ::
+  clientKeys <- replicateM nClients arbitrary
+  -- Prepare funding transaction which will give every client's
+  -- 'externalSigningKey' "some" lovelace. The internal 'signingKey' will get
+  -- funded before the benchmar run.
```
```suggestion
  -- funded in the beginning of the benchmark run.
```

ch1bo commented on hydra-cluster/src/HydraNode.hs:
```diff
@@ -161,23 +162,50 @@ waitForAll tracer delay nodes expected = do
+  response <- failAfter 3 $ queryNode hydraNodeId
+  let respBody = Req.responseBody response
+  when (Req.responseStatusCode response /= 200) $
+    failure ("Request for Hydra-node metrics failed :" <> show respBody)
```
This will not be called as `runReq` seemingly will throw a `HttpException` on errors.

Should: use `try` around `runReq` to provide more context.

ch1bo replied:
https://github.com/input-output-hk/hydra/commit/004317896a8b4567e48baeaf495c058c64ceb321

ch1bo commented on hydra-cluster/src/HydraNode.hs:
```diff
@@ -161,23 +162,50 @@ waitForAll tracer delay nodes expected = do
-    (HttpExceptionRequest _ (ConnectionFailure _)) -> threadDelay 100_000 >> cont
-    e -> throwIO e
+    (Req.VanillaHttpException _) -> threadDelay 100_000 >> cont
+    (Req.JsonHttpException _) -> threadDelay 100_000 >> cont
```
This will loop forever in case the request is wrong. i.e. these are not only connection exceptions. I'd remove it and see if it works consistently in tests. If not, then we re-introduce connection retries here by matching on the `HttpException (HttpExceptionRequest _ (ConnectionTimeout))` or so

ch1bo replied:
https://github.com/input-output-hk/hydra/commit/004317896a8b4567e48baeaf495c058c64ceb321

ch1bo commented on hydra-cluster/test/Test/EndToEndSpec.hs:
```diff
@@ -177,11 +178,15 @@ spec = around showLogsOnFailure $ do
-                send n3 $ input "Commit" ["utxo" .= Object mempty]
+                (aliceExternalVk, aliceExternalSk) <- generate genKeyPair
+                committedUTxOByAlice <- seedFromFaucet node aliceExternalVk aliceCommittedToHead Normal (contramap FromFaucet tracer)
+                (bobExternalVk, bobExternalSk) <- generate genKeyPair
```
Should: group bobs key with seeding and commiting for bob

ch1bo replied:
https://github.com/input-output-hk/hydra/commit/9e80f17f7f8da639d78b9dd60fff289b94d162db

ch1bo commented on hydra-node/json-schemas/api.yaml:
```diff
@@ -1098,19 +1099,7 @@ components:
-          properties:
-            tag:
-              type: string
-              enum: ["NoFuelUTXOFound"]
```
Should: keep the same errors as they still would exist, but only the "fuel marking" is different now. i.e. we just interpret everything held by the internal wallet as fuel

ch1bo replied:
https://github.com/input-output-hk/hydra/commit/3cd5abf930e8562d17f28f77b70075e18fa1ecfc

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Handlers.hs:
```diff
@@ -165,10 +165,8 @@ finalizeTx ::
-    Left ErrNoFuelUTxOFound ->
-      throwIO (NoFuelUTXOFound :: PostTxError Tx)
     Left ErrNotEnoughFunds{} ->
-      throwIO (NotEnoughFuel :: PostTxError Tx)
```
Should: keep the same errors (see above)

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Wallet.hs:
```diff
@@ -297,9 +288,14 @@ coverFee_ pparams systemStart epochInfo lookupUTxO walletUTxO partialTx@Babbage.
+          ins = traverse resolveInput (fst <$> utxoAsList)
+          totalIns = Coin $ sum $ either (const [0]) (fmap (unCoin . getAdaValue)) ins
+          totalOuts = foldMap getAdaValue (snd <$> utxoAsList)
+       in Left (ErrNotEnoughFunds ChangeError{inputBalance = totalIns, outputBalance = totalOuts})
```
Should: just keep the `ErrNoFuelUTxOFound` error.

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Wallet.hs:
```diff
@@ -365,16 +361,29 @@ coverFee_ pparams systemStart epochInfo lookupUTxO walletUTxO partialTx@Babbage.
+          [] -> Nothing
+          as ->
+            -- NOTE: here we are picking the **LAST** entry we found for no
+            -- particular reason.
```
It's not "no particular reason", but it's contrived. Seems like `gettriples'` gives us the lovelace as integer on the fst position and hence this is using the ledger api to sort by lovelace. So the last would be the biggest marked utxo here.

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Wallet.hs:
```diff
@@ -365,16 +361,29 @@ coverFee_ pparams systemStart epochInfo lookupUTxO walletUTxO partialTx@Babbage.
+findFuelOrLargestUTxO utxo =
+  case findFuelUTxO utxo of
+    Nothing ->
+      case UTxO.pairs $ UTxO.maxLovelaceUTxO apiUtxo of
```
Could: refactor this a bit to use alternative

i.e. something like
```
findFuelUTxO <|> findLargestUTxO
```

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/WalletSpec.hs:
```diff
@@ -227,6 +232,37 @@ isBalanced utxo originalTx balancedTx =
+        Nothing ->
+          property False
+            & counterexample ("Lookup UTXO: \n" <> decodeUtf8 (encodePretty lookupUTxO))
+            & counterexample ("Fuel UTXO: \n" <> decodeUtf8 (encodePretty fuelUTxO))
```
Could: Instead of doing the `counterexample` + `forAllBlind`, you could use `forAllShow` with these nicer strings you create here.

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/WalletSpec.hs:
```diff
@@ -227,6 +232,37 @@ isBalanced utxo originalTx balancedTx =
+          property False
+            & counterexample ("Lookup UTXO: \n" <> decodeUtf8 (encodePretty lookupUTxO))
+            & counterexample ("Fuel UTXO: \n" <> decodeUtf8 (encodePretty fuelUTxO))
+        Just (txin, txout) -> property $ Map.singleton txin txout == fuelUTxO
```
Could: use `===`

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/WalletSpec.hs:
```diff
@@ -227,6 +232,37 @@ isBalanced utxo originalTx balancedTx =
 
+prop_picksCorrectUTxOToPayTheFees :: Property
+prop_picksCorrectUTxOToPayTheFees =
+  forAllBlind genUTxO $ \lookupUTxO ->
```
Not sure why you would call this lookup utxo? There is nothing to lookup here?

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/WalletSpec.hs:
```diff
@@ -227,6 +232,37 @@ isBalanced utxo originalTx balancedTx =
+         where
+          fromLedgerUTxO = bimap Api.fromLedgerTxIn Api.fromLedgerTxOut
+          largestUTxO = UTxO.maxLovelaceUTxO $ UTxO.fromPairs $ fromLedgerUTxO <$> Map.toList combinedUtxO
+          expectedLargestUTxO = UTxO.singleton $ fromLedgerUTxO (txin, txout)
```
Should: not re-use the function which finds the max utxo here, instead we should rely on a property we know must hold after selection. For example: `txOutValue txout >= utxo1 && txOutValue txout >= utxo2`

ch1bo reviewed (changes requested):
In general looks good, but partial functions in a library is something should avoid strongly. Also there are several "should-do's".

ffakenz committed:
Refactor max utxo lovelance to not be partial

Also move the function to inner where clause where is needed.

This is also updating the property on finding the largest UTxO and
surprisingly it does not pass (even on the previous version) .

ch1bo committed:
Update hydra-cluster/src/Hydra/Cluster/Scenarios.hs

Make a comment a haddock

ch1bo committed:
Move signTx function into Hydra.Cardano.Api.Tx

ch1bo committed:
Undo removal of fuel errors

This keeps the same errors as they were originally. We still call the
funds on the internal wallet fuel, but we just don't require marking
them anymore.

ch1bo committed:
Use external commits in all but one e2e tests

ch1bo committed:
Do not loop on errors of getMetrics

As we are using req now, we can try to get rid of the retrying on
connection failures.

Also drops the dependency to http-conduit as currently unused.

ch1bo committed:
Update documentation of quickstart

Also remove details about fuel marking from demo instructions

ch1bo committed:
Removed more references of marking fuel

It's easier to not talk about details in all these tutorials, but just
refer to "fueling the hydra node".

ch1bo reviewed (approved):
Incorporated my change requests. Can approve now.

ch1bo committed:
Fix links in marked fuel deprecation warning
